File: commonutils-jre/src/main/java/net/maxsmr/commonutils/gson/PropertyBasedAdapter.java
package net.maxsmr.commonutils.gson;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import java.lang.reflect.Type;

import static net.maxsmr.commonutils.GsonUtilsKt.getJsonElement;
import static net.maxsmr.commonutils.GsonUtilsKt.getJsonPrimitive;

public class PropertyBasedAdapter<T> implements JsonSerializer<T>, JsonDeserializer<T> {

    private static final String KEY_SOURCE = "SOURCE";
    private static final String KEY_CLASS_META = "CLASS_META";

    @Override
    public JsonElement serialize(T src, Type typeOfSrc, JsonSerializationContext context) {
        JsonElement jsonElem = context.serialize(src, src.getClass());
        JsonObject object = new JsonObject();
        object.add(KEY_SOURCE, jsonElem);
        object.addProperty(KEY_CLASS_META, src.getClass().getCanonicalName());
        return object;
    }

    @Override
    public T deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        String className = getJsonPrimitive(json, KEY_CLASS_META, String.class);
        if (className != null) {
            Class<?> clz;
            try {
                clz = Class.forName(className);
            } catch (ClassNotFoundException e) {
                throw new JsonParseException(e);
            }
            return context.deserialize(getJsonElement(json, KEY_SOURCE, JsonElement.class), clz);
        }
        throw new JsonParseException(KEY_CLASS_META + " not specified");
    }
}

File: commonutils-jre/src/main/kotlin/net/maxsmr/commonutils/gson/PropertyBasedAdapter.kt
package net.maxsmr.commonutils.gson

import com.google.gson.*
import net.maxsmr.commonutils.getJsonElement
import net.maxsmr.commonutils.getJsonPrimitive
import java.lang.reflect.Type

/**
 * Регистрировать для абстрактных типов;
 * при сериализации происходит дописывание его имени класса,
 * десериализация делегируется для конкретного известного класса
 */
class PropertyBasedAdapter<T: Any> : JsonSerializer<T>, JsonDeserializer<T> {

    override fun serialize(src: T, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val jsonElement = context.serialize(src, src.javaClass)
        val obj = JsonObject()
        obj.add(KEY_SOURCE, jsonElement)
        obj.addProperty(KEY_CLASS_META, src.javaClass.name)
        return obj
    }

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): T {
        val className: String? = getJsonPrimitive(json, KEY_CLASS_META, String::class.java)
        className?.let  {
            val clz: Class<*> = try {
                Class.forName(className)
            } catch (e: ClassNotFoundException) {
                throw JsonParseException(e)
            }
            return context.deserialize(getJsonElement(json, KEY_SOURCE, JsonElement::class.java), clz)
        }
        throw JsonParseException("$KEY_CLASS_META not specified")
    }

    companion object {

        private const val KEY_SOURCE = "SOURCE"
        private const val KEY_CLASS_META = "CLASS_META"
    }
}